### DISCUSSION TITLE
UDP请求设置TOS要怎么修改？

### DISCUSSION THREAD

**Discussion by dev4u** [database64128/shadowsocks-go#37] [Q&A]
我想在这个项目中，对服务端发出的请求设置TOS，以提高使用体验。

目前情况是，tcp请求已经能在wireshark抓包中，看到dscp已经设置(大概在conn_linux)。但是udp，我一直没有找到合适的切入点。主要是我对这个项目的了解并不太深入。

所以想请教一下，我要对UDP做DSCP设置，大概要在哪块代码(文件)入手呢？我猜你也没太多时间，能领一下我找重点就行。

**Comment by database64128** [comment]
I'd like to implement this myself. The number of functions in `conn_linux.go` is already getting out of hand. Just give me a few hours. I'm going to clean it up a bit, then implement this feature for Linux, FreeBSD and macOS (including iOS and friends).

**Reply by dev4u** [reply]
好的，那我等你的消息。

**Reply by database64128** [reply]
7270b5fd4bd26c5704d69c343efea382a8690efc did a major refactoring of the `conn` package.

129094d3801862d0696bf6fffeeb618f575588a7 added support for traffic class.

**Reply by dev4u** [reply]
> 7270b5fd4bd26c5704d69c343efea382a8690efc did a major refactoring of the `conn` package.
> 
> 129094d3801862d0696bf6fffeeb618f575588a7 added support for traffic class.

好的！找时间我试试。提点小建议：如果设置的TOS值，是enum固定值让用户直接输入，会较容易使用。例如输入AF11、AF12、AF13、AF21……

![IMG_20230305_173522.jpg](https://user-images.githubusercontent.com/1210047/222952880-8b5dddfd-cee6-42cb-886f-1b0b0a24f640.jpg)

毕竟dscp不是那么好懂。譬如EF，十进制46，二进制101110。但是setsockopt需要传入184(10111000)才对应EF的TOS值。

虽然输入AF11这种也不好读懂，但至少不会设置错值。

**Reply by database64128** [reply]
https://linuxreviews.org/Type_of_Service_(ToS)_and_DSCP_Values

TL;DR: Use the value in the **ToS Decimal** column.

Only the first 6 bits are for DiffServ; the remaining 2 bits are used for ECN. Not that it's useful in any way, but I decided not to take away the possibility of specifying the ECN bits.

**Reply by dev4u** [reply]
好，我没意见。

**Reply by dev4u** [reply]
```log
2023-03-15T00:57:56.379+0800 WARN service/udp_nat.go:346 Failed to create UDP socket for new NAT session {"server": "socks5", "client": "ss-2022", "listenAddress": ":1081", "clientAddress": "[::ffff:127.0.0.1]:57145", "targetAddress": "220.181.174.226:443", "error": "listen udp :0: failed to set socket option IP_TOS: invalid argument"}
2023-03-15T00:57:56.381+0800 INFO service/tcp.go:272 Two-way relay started {"server": "socks5", "client": "ss-2022", "listenAddress": ":1081", "clientAddress": "[::ffff:127.0.0.1]:50575", "targetAddress": "clientservices.googleapis.com:443", "username": "", "initialPayloadLength": 1049}
2023-03-15T00:57:56.693+0800 WARN service/udp_nat.go:346 Failed to create UDP socket for new NAT session {"server": "socks5", "client": "ss-2022", "listenAddress": ":1081", "clientAddress": "[::ffff:127.0.0.1]:57145", "targetAddress": "220.181.174.226:443", "error": "listen udp :0: failed to set socket option IP_TOS: invalid argument"}
2023-03-15T00:57:57.305+0800 WARN service/udp_nat.go:346 Failed to create UDP socket for new NAT session {"server": "socks5", "client": "ss-2022", "listenAddress": ":1081", "clientAddress": "[::ffff:127.0.0.1]:57145", "targetAddress": "220.181.174.226:443", "error": "listen udp :0: failed to set socket option IP_TOS: invalid argument"}
2023-03-15T00:57:58.525+0800 WARN service/udp_nat.go:346 Failed to create UDP socket for new NAT session {"server": "socks5", "client": "ss-2022", "listenAddress": ":1081", "clientAddress": "[::ffff:127.0.0.1]:57145", "targetAddress": "220.181.174.226:443", "error": "listen udp :0: failed to set socket option IP_TOS: invalid argument"}
2023-03-15T00:58:00.389+0800 WARN service/udp_nat.go:346 Failed to create UDP socket for new NAT session {"server": "socks5", "client": "ss-2022", "listenAddress": ":1081", "clientAddress": "[::ffff:127.0.0.1]:57145", "targetAddress": "220.181.174.226:443", "error": "listen udp :0: failed to set socket option IP_TOS: invalid argument"}
```
本地运行的配置：
```json
{
  "servers": [
    {
      "name": "socks5",
      "listen": ":1081",
      "protocol": "socks5",
      "enableTCP": true,
      "listenerTFO": true,
      "enableUDP": true,
      "mtu": 1500,
      "natTimeoutSec": 180,
      "udpRelayBatchSize": 64,
      "udpServerRecvBatchSize": 1024,
      "udpSendChannelCapacity": 1024
    }
  ],
  "clients": [
    {
      "name": "ss-2022",
      "endpoint": "**********",
      "protocol": "2022-blake3-aes-128-gcm",
      "dialerTrafficClass": 184,
      "enableTCP": true,
      "dialerTFO": false,
      "enableUDP": true,
      "mtu": 1492,
      "psk": "********",
      "iPSKs": [
        "*******"
      ],
      "paddingPolicy": ""
    }
  ],
  "router": {
    "defaultTCPClientName": "ss-2022",
    "defaultUDPClientName": "ss-2022"
  }
}
```

**Reply by dev4u** [reply]
@database64128 麻烦看看这个异常，是不是我配置的问题？

**Reply by database64128** [reply]
Can you check if 862a46c6aea62d6f6d89efefae53459d447f666c fixed it?

**Reply by database64128** [reply]
Also make sure to do a packet capture to see if outgoing packets actually carry your configured ToS. You are running this on macOS, right? I only have Linux systems to test on.

**Reply by dev4u** [reply]
> Also make sure to do a packet capture to see if outgoing packets actually carry your configured ToS. You are running this on macOS, right? I only have Linux systems to test on.

是的，我的程序运行在macOS下。我的网络请求会先通过clashx做分流，需要代理的流量才会经过shadowsocks-go。

我回去会先抓包看看，是不是所有的UDP代理请求，都没带上TOS。

**Reply by dev4u** [reply]
> Can you check if [862a46c](https://github.com/database64128/shadowsocks-go/commit/862a46c6aea62d6f6d89efefae53459d447f666c) fixed it?

@database64128 更新后，确认没有再发生`failed to set socket option IP_TOS: invalid argument`异常。

更新代码前，我也抓过包，所有代理的udp请求，都没成功设置TOS。

**Reply by database64128** [reply]
Without this fix, it fails to set the socket option and closes the socket immediately, so you won't see any outgoing packets.

What I need to know is, with the fix, do outgoing IPv4 packets have the ToS bits set?

**Reply by database64128** [reply]
The problem with macOS, and possibly many other BSDs, is that they only support the bare minimum of dual-stack sockets. Unlike Linux and Windows, these BSDs do not allow setting IPv4 socket options on dual-stack IPv6 sockets.

Currently, shadowsocks-go uses a single dual-stack socket for outgoing packets in each session. I do have plans to support using separate sockets for v4 and v6, in order to support custom bind addresses, but it'll take some time to implement. And since the problem does not affect Linux or Windows, it is fairly low priority for me.

For now, if you want traffic class to work for both v4 and v6, use Linux. Or you can make a small temporary change to the version running on your macOS device: Change `"udp"` to `"udp4"` in the following line.

https://github.com/database64128/shadowsocks-go/blob/12ea09d760894e049d2cc165cf06a24841bc5d58/service/udp_nat.go#L344

Note that with this change UDP will no longer work for IPv6 endpoints. But it should be sufficient for your use case.

**Reply by dev4u** [reply]
如果你问的是tcp4有没带上TOS，答案是有带的，但udp4还是没有。
我在wireahark里，设置了过滤条件：ip.dsfield.dscp != 0，但没有命中udp协议数据包被显示。
![MVIMG_20230316_024610.jpg](https://user-images.githubusercontent.com/1210047/225412386-83cc8d98-fd24-4a12-bb01-0ad2cc0002b9.jpg)
